src/GNDStk/Node.hpp

```diff
@@ -0,0 +1,187 @@
+
+#include "GNDStk/Node/src/detail.hpp"
+
+// This is returned by some of Node's child() functions
+template<class NODE, class T>
+class TypedNode;
+
+
+
+// -----------------------------------------------------------------------------
+// Node (templated)
+// node = Node<>
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class METADATA_CONTAINER, // container type for metadata
+   template<class...> class CHILDREN_CONTAINER  // container type for children
+>
+class Node {
+   using metaPair = std::pair<std::string,std::string>;
+   using childPtr = std::unique_ptr<Node>;
+
+public:
+
+   // ------------------------
+   // Data
+   // ------------------------
+
+   // Simple node for our tree structure:
+   //    name
+   //    metadata (container of string pairs)
+   //    children (container of pointers to other Node<>s)
+   std::string name;
+   METADATA_CONTAINER<metaPair,std::allocator<metaPair>> metadata;
+   CHILDREN_CONTAINER<childPtr,std::allocator<childPtr>> children;
+
+   // ------------------------
+   // Simple functions
+   // ------------------------
+
+   // clear()
+   Node &clear()
+   {
+      name = "";
+      metadata.clear();
+      children.clear();
+      return *this;
+   }
+
+   // is this node empty?
+   bool empty() const
+   {
+      return name == "" && metadata.size() == 0 && children.size() == 0;
+   }
+
+   // is this a leaf node?
+   bool leaf() const
+   {
+      return children.size() == 0;
+   }
+
+   // copy
+   Node copy() const
```

Comment:
Good question, with what I hope is a good answer. Deleting a major object's copy constructor, and its assignment operator, isn't something I'd normally do. Depending on how my concerns pan out, and how the experiences of initial users pan out, there's a non-trivial chance that I'll change this back, and just do things in the standard way. The standard way has advantages - like in the fact that it is, well, standard. Expected, at least by more-knowledgeable users. If I do change it, the changes should be straightforward to do, and with no (or very little) breaking of user codes.

My thought was as follows. What are the likely uses of GNDStk? I'd guess that users will typically be doing one or more of:

   * Reading and querying a GNDS tree.

   * Reading the tree, augmenting or otherwise tweaking it (say, to fix incorrect data), then writing it back out.

   * Creating a tree from scratch, then writing it.

A common denominator in all of those activites is that someone is working inside of a tree, at least in terms of where the data are being read from, or written to.

So, imagine this scenario: a non-expert in C++ writes something like the following to extract a node:

```
   auto node = atree(something that pulls out a node);
   // or:
   Node<> node = atree(something that pulls out a node);
```

Then they modify node, write out the tree, and...wonder why nothing changed. Or, worse, they don't notice that nothing changed until they hear back from the person who asked them to fix it! You and I can see the error, but will a typical user see it?

One could also imagine a user doing lots of things like the above, especially with a higher-up node in a large GNDS hierarchy, and wondering why the code is so slow, or why so much memory is being used.

So, given the typically ways in which I'm guessing most users will want to be interacting with the data, I decided I'd force the issue (basically at compile time) of doing it right, and making someone *explicitly* write .copy() if they really want a copy.

Of course, the counterargument would be that it's completely normal, in normal C++ codes, for someone to have the usual choice of either dealing with an object in its original place, or getting a copy. Imagine making a variable that gets, say, some_vector[n]. With this in mind, yes, perhaps I shouldn't be trying to do anything differently. Which brings me back to: I'd like to see how this pans out for actual GNDStk users. (Note: for whatever reasons, whether similar to mine or completely different, pugi::xml_document's copy constructor and copy assignment operator would appear to be private. I haven't looked around and tried to determine what the author's reasoning was for that.)

Reply:
My preference would be to use the standard C++ constructors do this. For the GNDStk library, I'm not interested in pandering to those with little to no C++ experience. GNDStk is intended to be a low-level API. We will build applications with GNDStk that could simplify things for the less-informed end user.

Reply:
With tests for tree and node being pretty much filled out now, I have a note to go in and try putting back the copy constructor, and making sure (via tests) that the sky doesn't fall. With regards to GNDS being a low-level API: that's great, and my hope is that it serves well as just that. My hope also is that it's easy and intuitive enough to use that even a not-too-sophisticated C++ programmer will find it straightforward and pleasant to use directly, if they wish to. Only time will tell if I've achieved that.

Reply:
In fact it was a while back that I went ahead and put in a copy constructor for Node. At the time I left the copy() in there; and it looks like I still use it in a few places. I recall that I left it in for auditing purposes - I wanted to get a good handle on just when and where copying is needed, in order to reason about efficiency. An explicit copy() call is easier to discern, in the source, from a (typically implicit) copy constructor. I'll make a note to go in and review all this again, and perhaps remove the old copy() entirely.